### PR TITLE
Call operation performance optimizations 

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

- Introduce "undoable" journoled collections: set, map, table, and list
- Move values that are the same or shared across the transaction to
  TxValues record
- Move warm and cold storage to undoable sets and tables.
- Move transient storage to undoable table.
- Move address hashing inside of address with a memoized field.
- lazy create EOF return stack

### CODE REVIEW

**Comment by shemnon** [comment]
8%->80% performance improvement in the call/delegate call operation, respecitivly.

call
```
evm --repeat
1000
--code
6113376102005260005b61047c81101561050a5760206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af15060206101e0602061020060007372657475726e207465737420636f6e74726163745af150600181019050610009565b50
```
pre
Screenshot 2023-06-05 at 11 48 59 AM
post
Screenshot 2023-06-05 at 11 44 48 AM

delegate call
```
evm --repeat 1000 --code 61e11e6102805260005b61043e81101561050857602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450602061026060006102807372657475726e207465737420636f6e74726163745af450600181019050610009565b50
```
pre
Screenshot 2023-06-05 at 11 59 31 AM
post
Screenshot 2023-06-05 at 12 02 07 PM

File: datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java [hunk 1]
```diff
@@ -75,6 +80,16 @@ public class Address extends DelegatingBytes {
   /** The constant ZERO. */
   public static final Address ZERO = Address.fromHexString("0x0");
 
+  static LoadingCache<Address, Hash> hashCache =
```
Comment (matkt) [inline datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java:83]: we don't   want to add a limit for this cache ?

File: evm/src/test/java/org/hyperledger/besu/collections/undo/UndoListTest.java [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    assertThat(subject.mark()).isGreaterThan(mark);
+
+    mark = subject.mark();
+    // non-changing undo does not advance mark
+    subject.undo(mark);
+    assertThat(subject.mark()).isEqualTo(mark);
+
+    mark = subject.mark();
+    // non-existent remove doesn't advance mark
+    subject.remove("Bonjour");
```
Comment (matkt) [inline evm/src/test/java/org/hyperledger/besu/collections/undo/UndoListTest.java:76]: 🇫🇷

File: evm/src/test/java/org/hyperledger/besu/evm/toy/ToyAccount.java [hunk 3]
```diff
@@ -40,7 +40,7 @@ public class ToyAccount implements EvmAccount, MutableAccount {
 
   private Address address;
   private final Supplier<Hash> addressHash =
-      Suppliers.memoize(() -> address == null ? Hash.ZERO : Hash.hash(address));
+      Suppliers.memoize(() -> address == null ? Hash.ZERO : address.addressHash());
```
Comment (matkt) [inline evm/src/test/java/org/hyperledger/besu/evm/toy/ToyAccount.java:43]: is it still need to have memoize in this case ?

**Comment by ahamlat** [comment]
> 8%->80% performance improvement in the call/delegate call operation, respecitivly.

When I look to the CPU profiling in case of Call operation, I see the roughly the same percentage in samples pre and post this PR, around 63%. Is it the right screenshot ? Am I missing something ?

File: datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java (hunk 1 above)
Comment (shemnon) [reply datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java:83]: What limit would you suggest?

File: evm/src/test/java/org/hyperledger/besu/evm/toy/ToyAccount.java (hunk 3 above)
Comment (shemnon) [reply evm/src/test/java/org/hyperledger/besu/evm/toy/ToyAccount.java:43]: I applied it to all cases where we saw `Hash.hsah(addres))` to be consistant.

**Comment by shemnon** [comment]
Assuming sampling is at the same fixed interval, the CALL samples had 3190 pre and 2943, so pre had 8.39% more samples.  for DELEGATECALL the change was more dramantic at 4152/2249, so pre had 84.61% more samples.  I'll get my CartEVM results tomorrow.

I also don't have a good "tower" call benchmark (where we do 1000 deep callstacks).  I'll dig into reference tests for a good example.  That is where I expect the most payback for undo collections.

What surprised me is just how much time is spent in the call frame init as opposed to other work steps.

**Comment by shemnon** [comment]
I also want at a day of fuzzing before it gets committed.

- [ ] Full day of fuzzing.

File: datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java (hunk 1 above)
Comment (matkt) [reply datatypes/src/main/java/org/hyperledger/besu/datatypes/Address.java:83]: I would say a maximum size. Maybe we should see how much on average we load hash by block and got a number. But with a cache without max seems dangerous to me . Or else there is a default limit?

**Comment by shemnon** [comment]
Tower call results are unimpressive, so given the breadth of the change I'm going to shelve it as not worth the risk.
Only 10% on calls...

```
./ethereum/evmtool/build/install/evmtool/bin/evm state-test ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stDelegatecallTestHomestead/Delegatecall1024.json
```

pre
```
{"output":"","gasUsed":"0x31e845","time":18264833,"Mgps":"179.072","test":"Delegatecall1024","fork":"Istanbul","d":0,"g":0,"v":0,"postHash":"0x5e34e5c4b1eec1379e8c442fcf05a38a557c319a5a0904c62e4a9071f9f5acbc","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x7c335","time":18224834,"Mgps":"27.914","test":"Delegatecall1024","fork":"London","d":0,"g":0,"v":0,"postHash":"0xfc071ef580274c1c82c28f60253849317c156f6159f711582ef93da01f26f884","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x7c335","time":18216584,"Mgps":"27.926","test":"Delegatecall1024","fork":"Merge","d":0,"g":0,"v":0,"postHash":"0xfc071ef580274c1c82c28f60253849317c156f6159f711582ef93da01f26f884","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x7c335","time":18320208,"Mgps":"27.769","test":"Delegatecall1024","fork":"Shanghai","d":0,"g":0,"v":0,"postHash":"0xfc071ef580274c1c82c28f60253849317c156f6159f711582ef93da01f26f884","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
```

post
```
{"output":"","gasUsed":"0x31e845","time":16720084,"Mgps":"195.617","test":"Delegatecall1024","fork":"Istanbul","d":0,"g":0,"v":0,"postHash":"0x5e34e5c4b1eec1379e8c442fcf05a38a557c319a5a0904c62e4a9071f9f5acbc","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x7c335","time":17400291,"Mgps":"29.237","test":"Delegatecall1024","fork":"London","d":0,"g":0,"v":0,"postHash":"0xfc071ef580274c1c82c28f60253849317c156f6159f711582ef93da01f26f884","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x7c335","time":16167792,"Mgps":"31.465","test":"Delegatecall1024","fork":"Merge","d":0,"g":0,"v":0,"postHash":"0xfc071ef580274c1c82c28f60253849317c156f6159f711582ef93da01f26f884","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x7c335","time":16568500,"Mgps":"30.704","test":"Delegatecall1024","fork":"Shanghai","d":0,"g":0,"v":0,"postHash":"0xfc071ef580274c1c82c28f60253849317c156f6159f711582ef93da01f26f884","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
```

Berlin is where access lists were introduced and is where the showdown is.

If actual on-chain gas limits are used (due to 63/64) stats are _much_ better, but only a 70 high tower gets built

```
{"output":"","gasUsed":"0x233526","time":6176625,"Mgps":"373.564","test":"Delegatecall1024OOG","fork":"Istanbul","d":0,"g":0,"v":0,"postHash":"0x803f3435d98e5077d734e9cc87f1c7473634a9b9a12849894ea0231f9c4287fd","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x1aabcf","time":6702500,"Mgps":"260.786","test":"Delegatecall1024OOG","fork":"London","d":0,"g":0,"v":0,"postHash":"0xf0a9d54c9e9ae8a9500af004bc5a4bfaa136fef8b56f105d9db0530dedd23051","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x1aabcf","time":5973000,"Mgps":"292.637","test":"Delegatecall1024OOG","fork":"Merge","d":0,"g":0,"v":0,"postHash":"0xf0a9d54c9e9ae8a9500af004bc5a4bfaa136fef8b56f105d9db0530dedd23051","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"","gasUsed":"0x1aabcf","time":6280500,"Mgps":"278.309","test":"Delegatecall1024OOG","fork":"Shanghai","d":0,"g":0,"v":0,"postHash":"0xf0a9d54c9e9ae8a9500af004bc5a4bfaa136fef8b56f105d9db0530dedd23051","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
```

**Comment by shemnon** [comment]
[Performance numbers](https://docs.google.com/spreadsheets/d/1xd7PXWaQz3VWqP2UN6zLLZ1sa2Y1B7W1Grq4Op9FjKI/edit?usp=sharing)

* "all in" performance is a small drop within statistical noise. [data](https://docs.google.com/spreadsheets/d/1xd7PXWaQz3VWqP2UN6zLLZ1sa2Y1B7W1Grq4Op9FjKI/edit#gid=349651545&range=R159)
* performacne of CALL, DELEGATECALL, and STATICCALL improve 20%-26% [data](https://docs.google.com/spreadsheets/d/1xd7PXWaQz3VWqP2UN6zLLZ1sa2Y1B7W1Grq4Op9FjKI/edit#gid=349651545&range=R146:R150)
* Selfdestruct takes a nearly 80% hit, but that is from an absurd 5.9 GigaGas/sec to 1.3 GigaGas/sec [data](https://docs.google.com/spreadsheets/d/1xd7PXWaQz3VWqP2UN6zLLZ1sa2Y1B7W1Grq4Op9FjKI/edit#gid=349651545&range=L152:R152)

Call ops are still the worst performers, so raising the floor is worth the hit to a lesser used top performing operation.

**Comment by shemnon** [comment]
A week of fuzzing had no findings
```
INFO [08-07|12:48:59.102] Executing                                tests=6,739,367 time=156h0m40.003s test/s=12.0 "avg steps"=4310.0 global=6,961,318
```

**Review by daniellehrner — APPROVED** [review]
LGTM

**Pull request merged**